Don't introduce `arg` as-pattern if choice has `arg` field

akrmn opened the pull request:
This PR grew a bit beyond its original scope, so I recommend reviewing it commit-by-commit

Part of https://github.com/digital-asset/daml/issues/3017

Daml PR: https://github.com/digital-asset/daml/pull/13755

akrmn committed:
Drop uses of XPat constructor

The XPat constructor doesn't add any information, we can just use the underlying
Pat constructor directly. According to the commit message of (upstream)
182b119943d34e82f67525c4b2390557f060c5f9, XPat was only used to attach source
locations, but in our uses the location was always 'noLoc'.

remyhaemmerle-da reviewed (approved):
Look reasonable. Thanks.

akrmn committed:
Use viEnsure binding in validateInterface

akrmn committed:
Clean up checkNumArgs

akrmn committed:
Remove unused top level bindings in RdrHsSyn

akrmn committed:
Move Pat helpers together

akrmn committed:
Move 'this' pattern definitions up the call chain

akrmn committed:
Don't bind 'arg' to choice record if it has 'arg' field

asPatRecWild now takes an argument for the shape of the type being matched, which allows omitting the '..' when the type has no field, and the '@'-bound variable if it clashes with a field

akrmn committed:
Don't bind 'arg' to choice record if template has 'arg' field

akrmn commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2437,28 +2437,57 @@ mkRdrPat = VarPat noExt . noLoc
+        AsPat noExt (noLoc $ mkRdrUnqual $ mkVarOcc varName) . noLoc
+
+-- | Utility for constructing patterns of the form 'T{..}',
+-- omitting the '..' for types without labelled fields.
```
this is a bit better than using `PrefixCon` for types without fields because `{}` also works for non-record product types, i.e.

```
data Prod = Prod Int Char

f : Prod -> Bool
f Prod {} = True
```

sofiafaro-da commented on compiler/parser/RdrHsSyn.hs:
```diff
@@ -2289,7 +2289,7 @@ validateInterface name requires decls = do
     , viFunctionSignatures = ibdFunctionSignatures
     , viChoices = ibdChoices
-    , viEnsure = listToMaybe ibdEnsures
+    , viEnsure
```
Nice catch!

akrmn replied:
to be fair it was my own bug :(

sofiafaro-da replied:
Still a good catch :) We all write bugs.

sofiafaro-da reviewed (approved):
Nice, thank you!

akrmn commented:
Closing in favor of just forbidding the problematic fields, see full explanation on https://github.com/digital-asset/daml/pull/13755
